### DISCUSSION TITLE
Multiple Sensors on Single SpanPoint

### DISCUSSION THREAD

**Discussion by cfosteriv** [HomeSpan/HomeSpan#875] [Q&A] [answered]
Have a a single SpanPoint that has two temp sensors and a humidity sensor (a dht22 and a DS18**). For ambient temp and internal temp/humidity. 

Would like to send these three readings back without needing to use 2 or 3 different esp32 modules as span points (excluding the main device).

For some reason I cannot link a humidity sensor to a temp sensor, or a temp sensor to another sensor, to utilize one accessory/SpanPoint to provide these 3 variables to the home app.

Right now my single spanupoint device and main device do send/receive the data as they should. I have created a struct and the data does pass SpanPoint->HomeSpan correctly. It's just the representation of the data when updating the home app that seems to not be working. 

I did notice that if I simple make two accessories utilizing the same SpanPoint class and provide the same MAC address, it does produce two temps in the Home App, one right next to each other. If you select it, it brings up a menu that shows both of the temps, however only one ever updates even though all the data is being sent back.  The one that updates is always the first accessory to be added. 

Wondering if there is a way to make a linked service that will work native with the Home app for the two temperatures and humidity sensors so they can operate on one SpanPoint device? Maybe I am misunderstanding or not implementing queuing in an effective way to accomplish this?

**Comment by HomeSpan** [comment, marked as answer]
There is not a one-to-one relationship between SpanPoint and HomeKit Accessories - they are completely independent.   Also, Apple does not support any single Services that contains multiple standalone temp/humidity sensors.  You will need to define 3 separate Accessories (all within the same Main ESP32) to support two temp sensors and one humidity sensor connected to the Remote ESP32.

Note the Home App groups sensors of the same type together for the purposes of displaying on the Home App pages.  I frankly find this annoying but Apple made this change quite a while ago so we are sort of stuck with it.  However, as you found, by clicking the temp icon at the top of the Home App you can see all the underlying sensors.  Also, if each sensor is in a different room, by selecting a view for just that room the Home App icon will show just the standalone sensors assigned to that room.

For the SpanPoint linkage, I recommend using a single structure to pass all three readings at the same time.  On the Main Device, there are lots of ways you can create a single SpanPoint object, read all three variables at once, and then parse them out to update each of the three Accessories:

-  create a SpanPoint object within just one of the Accessories and when it updates, you can update the Accessory with one of the sensor readings and set global variables with the other two sensor readings to get picked up by them in their own Service loops;

- create a global SpanPoint object and have the reading occur in the main Arduino loop.  Within that loop you can then update all three Accessories when an update is found.  See the [External References Tutorial](https://github.com/HomeSpan/HomeSpan/blob/master/examples/Other%20Examples/ExternalReference/ExternalReference.ino) for a good example of how to do something like this.  I'd recommend going this route;

- create a custom Service that creates and maintains three other standalone sensor Services.  There is not much to be gained over this than simply creating a global SpanPoint object (as described above) but it would provide a little extra encapsulation of all the logic within a single Service structure (that will not be shown in the Home App) in case you wanted to add other Accessories to the Main Device that were unrelated to the Remote Device.  I have not tried this method, but it may work well in some circumstances.

**Reply by cfosteriv** [reply]
Ok so the double sensor data under a single 'Temperature' title is normal behavior, Actually had thought maybe it exposed a bug due to two values for the same room, of the same sensor, provided by the the same MAC addressed device.

**Reply by HomeSpan** [reply]
Yes, multiple temp sensors "hidden" under a single Temp Icon at the top of the Home App is common.   This icon will usually show a range of low-high temperatures across all the sensors included.   From the main Home App page, this includes all temperature sensors in your Home.  If you select a certain room, it will show the range for only those sensors in that room.
